Search: remove unexpected sidebar widgets added on theme switch

kangzj opened the pull request:
Partly fixes #18556

#### Changes proposed in this Pull Request:
Move widgets added by WP to inactive widgets when new theme has less sidebars than the old theme, by hooking `after_switch_theme` to save a copy of old sidebar setting and hooking `pre_update_option_sidebars_widgets` to do the sweeping.

Another possible solution is to change the instant search sidebar name from `jetpack-instant-search-sidebar` to a name without keyword `sidebar`. The hack would walk around the widgets migrations strategy by WordPress. The problem with its compatibility, which will result in some code migrating the old setting to the new one otherwise search sidebar settings would be lost.

#### Jetpack product discussion
pcNPJE-96-p2

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
- Go to a site with Jetpack Search subscription
- Switch theme to `Twenty Twenty` which has two sidebar area ("Footer #1" and "Footer #2"), make sure both sidebars have some widgets (you can manage widgets here: /wp-admin/widgets.php)
- Switch to a theme that has less sidebars, say Twenty Twenty-One which has only one sidebar ("Footer").
- Ensure that the Jetpack Search sidebar has the same widgets before the theme switch, meaning it doesn't have a bunch of unwanted widgets in it, like 'Archives', 'Categories' and 'Meta' etc.

kangzj committed:
[not verified] remove unexpected sidebar widgets added by WP on theme switch

kangzj committed:
[not verified] empty the saved sidebars_widgets value

kangzj committed:
[not verified] changelog

kangzj committed:
[not verified] added test_remove_wp_migrated_widgets

bluefuton commented on projects/plugins/jetpack/changelog/fix-search-sidebar-unexpected-widgets:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: bugfix
+
+Search: fixed standard sidebar widgets can end up in the search modal sidebar when switching themes
```
```suggestion
Search: prevent standard sidebar widgets ending up in the search modal sidebar when switching themes
```

jsnmoon commented on projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php:
```diff
@@ -43,6 +78,10 @@ public function init_hooks() {
 			add_action( 'update_option', array( $this, 'track_widget_updates' ), 10, 3 );
+			// The priority has to be lower than 10 to run before _wp_sidebars_changed.
+			// Which migrates widgets from old theme to the new one.
+			add_action( 'after_switch_theme', array( $this, 'save_old_sidebars_widgets' ), 5 );
```
The [official docs](https://developer.wordpress.org/reference/hooks/after_switch_theme/) indicate that this action is fired multiple times. Is this desirable? Would using [`switch_theme`](https://developer.wordpress.org/reference/hooks/switch_theme/) be more appropriate here?

kangzj replied:
Yep it might fire multiple times, it doesn't matter. If it fired once already, the process in `remove_wp_migrated_widgets` wouldn't run as the jetpack sidebar have the same size.

`switch_theme` is fired in a separate request before the one  that fires `after_switch_theme`, so the mechanism in the PR wouldn't work.

jsnmoon commented on projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php:
```diff
@@ -43,6 +78,10 @@ public function init_hooks() {
 			add_action( 'update_option', array( $this, 'track_widget_updates' ), 10, 3 );
+			// The priority has to be lower than 10 to run before _wp_sidebars_changed.
+			// Which migrates widgets from old theme to the new one.
+			add_action( 'after_switch_theme', array( $this, 'save_old_sidebars_widgets' ), 5 );
```
I'm guessing that `pre_update_option_sidebars_widgets` is fired after `after_switch_theme`. Is this correct?

kangzj replied:
Yep. The process detected theme change is hooked to `init` by default, it fires `after_switch_theme` when it detects theme change and it'll update theme options afterwards, which would trigger `pre_update_option_sidebars_widgets`.

If it doesn't detect theme change, then the whole thing wouldn't run at all.

jsnmoon reviewed (changes requested):
Unfortunately, this change behaved differently than what I expected during my testing. Here's what I did:

1) Activated the 2020 theme.
2) Added a Jetpack Search widget to the `Jetpack Search Sidebar`.
3) Added Search, Recent Posts, and Recent Comments widgets to `Footer 1`.
4) Added Archives, Categories, and Meta widgets to `Footer 2`.
5) Activated the 2021 theme.

What I expected:

- `Jetpack Search Sidebar` contains the Jetpack Search widget.
- `Footer 1` contains the Search, Recent Posts, and Recent Comments widgets.
- `Footer 2` is gone.
- Archives, Categories, and Meta widgets are moved to the Inactive Widgets section.

What I experienced:
- ⛔ `Jetpack Search Sidebar` contains no widgets.
- ✅ `Footer 1` contains the expected widgets.
- ✅ `Footer 2` is gone.
- ⛔ Archives, Categories, and Meta widgets have been permanently removed instead of being moved to the Inactive Widgets section.

kangzj commented:
> Unfortunately, this change behaved differently than what I expected during my testing. Here's what I did:
> 
> 1. Activated the 2020 theme.
> 2. Added a Jetpack Search widget to the `Jetpack Search Sidebar`.
> 3. Added Search, Recent Posts, and Recent Comments widgets to `Footer 1`.
> 4. Added Archives, Categories, and Meta widgets to `Footer 2`.
> 5. Activated the 2021 theme.
> 
> What I expected:
> 
> * `Jetpack Search Sidebar` contains the Jetpack Search widget.
> * `Footer 1` contains the Search, Recent Posts, and Recent Comments widgets.
> * `Footer 2` is gone.
> * Archives, Categories, and Meta widgets are moved to the Inactive Widgets section.
> 
> What I experienced:
> 
> * ⛔ `Jetpack Search Sidebar` contains no widgets.
> * ✅ `Footer 1` contains the expected widgets.
> * ✅ `Footer 2` is gone.
> * ⛔ Archives, Categories, and Meta widgets have been permanently removed instead of being moved to the Inactive Widgets section.

Thanks for spotting this @jsnmoon looking now

kangzj committed:
[not verified] fix parameter signature for after_switch_theme

kangzj commented:
Hi @jsnmoon sorry this is totally my bad. I added a parameter to function `save_old_sidebars_widgets` for the sake of unit testing and didn't regress. It's because actions passes in one parameter by default which is absolutely not the parameter the function needs.

Fixed in 3ef95f6

jsnmoon commented on projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php:
```diff
@@ -665,4 +704,46 @@ public function auto_config_woo_result_format() {
+
+		foreach ( $sidebars_widgets as $sidebar => $widgets ) {
+			if ( 0 === stripos( $sidebar, static::JETPACK_INSTANT_SEARCH_SIDEBAR ) ) {
+				if ( count( $sidebars_widgets[ static::JETPACK_INSTANT_SEARCH_SIDEBAR ] ) <= count( $this->old_sidebars_widgets[ static::JETPACK_INSTANT_SEARCH_SIDEBAR ] ) ) {
```
When would the count of widgets in the current sidebar be less than the widgets in the old sidebar?

kangzj replied:
```
// An 'equal' would work for the condition, but we want to cover 'less than' as well, just for extra insurance.
// If the new Jetpack sidebar has already got less widgets, no need to run the code following the condition.
```
Added comments in 4fcef02

jsnmoon commented on projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php:
```diff
@@ -12,6 +12,41 @@
+	const JETPACK_INSTANT_SEARCH_SIDEBAR = 'jetpack-instant-search-sidebar';
+
+	/**
+	 * Variable to save old sidebars_widgets value
```
Could we describe when this variable is saved/deleted? Describing the action sequence for `after_switch_theme` and `pre_update_option_sidebars_widgets` could be useful for future contributors.

kangzj replied:
```
The value is set when action `after_switch_theme` is applied and cleared on filter `pre_update_option_sidebars_widgets`.
The filters mentioned above run on /wp-admin/themes.php?activated=true, a request closely following switching theme.
```

Added comments in 4fcef02

jsnmoon reviewed (dismissed):
Excellent work tracking this down and adding tests to boot!

I left some questions/nits in the code below, but none of them are blockers. The fix works as described.

jsnmoon commented:
Oh, it looks like the `WP_Test_Jetpack_Instant_Search::test_remove_wp_migrated_widgets` test is failing with the message `Error: Call to undefined method WP_Test_Jetpack_Instant_Search::assertEqualsCanonicalizing()`.

I also recommend rebasing to pass the up-to-date check.

kangzj committed:
Added comments for future reference

kangzj committed:
Fix failing unit test

kangzj commented:
> Oh, it looks like the `WP_Test_Jetpack_Instant_Search::test_remove_wp_migrated_widgets` test is failing with the message `Error: Call to undefined method WP_Test_Jetpack_Instant_Search::assertEqualsCanonicalizing()`. We probably want to fix that before passing this along to Crew.
> 
> I also recommend rebasing the feature branch to pass the up-to-date check.

Ah, thanks for catching this! @jsnmoon. I didn't notice this and weird enough it passes on my local, probably because of phpunit versions? (unit tests run on docker and they should have same everything...)

Changed to a less fancy (older) assertion and rebased, let's see how it goes.

kangzj commented:
Hey @jsnmoon sorry for the many back and forth. All tests passed now, could you please have another look?

Commit:
Merge branch 'master' of github.com:Automattic/jetpack into fix/search-sidebar-unexpected-widgets

jsnmoon reviewed (dismissed):
👍 🚢 

Merged master into the branch to pass the `PR is up to date` check!

jeherve committed:
Merge remote-tracking branch 'origin/master' into fix/search-sidebar-unexpected-widgets

jeherve reviewed (approved):
This tests well for me. 🚢

jsnmoon commented:
Shipped in r226262-wpcom.
